packages/component-library/components/Heading/Heading.elm

```diff
@@ -0,0 +1,270 @@
+
+                Nothing ->
+                    inferTagFromVariant config.variant
+
+        resolveCustomClass =
+            case config.classNameAndIHaveSpokenToDST of
+                Just classNameAndIHaveSpokenToDST_ ->
+                    classNameAndIHaveSpokenToDST_
+
+                Nothing ->
+                    ""
+    in
+    resolveTag
+        ([ className config.variant ] ++ [ Html.Attributes.class resolveCustomClass ] ++ resolveId ++ resolveAttributes)
+        children
+
+
+className : TypeVariant -> Html.Attribute msg
+className typeVariant =
+    let
+        styleClass =
+            case typeVariant of
+                Display0 ->
+                    .display0
+
+                Heading1 ->
+                    .heading1
+
+                Heading2 ->
+                    .heading2
+
+                Heading3 ->
+                    .heading3
+
+                Heading4 ->
+                    .heading4
+
+                Heading5 ->
+                    .heading5
+
+                Heading6 ->
+                    .heading6
+    in
+    styles.classList
+        [ ( styleClass, True )
+        , ( .heading, True )
+        ]
+
+
+styles =
+    css "@kaizen/component-library/components/Heading/Heading.module.scss"
+        { heading = "heading"
+        , display0 = "display-0"
+        , heading1 = "heading-1"
+        , heading2 = "heading-2"
+        , heading3 = "heading-3"
+        , heading4 = "heading-4"
+        , heading5 = "heading-5"
+        , heading6 = "heading-6"
+        }
+
+
+type TypeVariant
+    = Display0
+    | Heading1
+    | Heading2
+    | Heading3
+    | Heading4
+    | Heading5
+    | Heading6
+
+
+
+-- MODIFIERS
+
+
+pre : Config msg
+pre =
+    Config { defaultConfig | tag = Just Html.pre }
+
+
+p : Config msg
+p =
+    Config { defaultConfig | tag = Just Html.p }
+
+
+a : Config msg
+a =
+    Config { defaultConfig | tag = Just Html.a }
+
+
+div : Config msg
+div =
+    Config { defaultConfig | tag = Just Html.div }
+
+
+span : Config msg
+span =
+    Config { defaultConfig | tag = Just Html.span }
+
+
+h1 : Config msg
+h1 =
+    Config { defaultConfig | tag = Just Html.h1 }
+
+
+h2 : Config msg
+h2 =
+    Config { defaultConfig | tag = Just Html.h2 }
+
+
+h3 : Config msg
+h3 =
+    Config { defaultConfig | tag = Just Html.h3 }
+
+
+h4 : Config msg
+h4 =
+    Config { defaultConfig | tag = Just Html.h4 }
+
+
+h5 : Config msg
+h5 =
+    Config { defaultConfig | tag = Just Html.h5 }
+
+
+h6 : Config msg
+h6 =
+    Config { defaultConfig | tag = Just Html.h6 }
+
+
+variant : TypeVariant -> Config msg -> Config msg
+variant value (Config config) =
+    Config { config | variant = value }
+
+
+id : String -> Config msg -> Config msg
+id id_ (Config config) =
+    Config { config | id = Just id_ }
+
+
+classNameAndIHaveSpokenToDST : String -> Config msg -> Config msg
+classNameAndIHaveSpokenToDST value (Config config) =
+    Config { config | classNameAndIHaveSpokenToDST = Just value }
+
+
```

Comment:
I think this should be named `addDataAttribute`, if you can only add `data-` attributes.
OR, the `DataAttribute` type should be renamed `Attribute` and allow for other attributes other than just `data-` attributes (even if for now data attribute is the only possible one).

Eg:

```
type Attribute =
    DataAttribute { key : String, value : String }
```
This is a custom type with only one option, but you could add other options later.

I don't mind which way we go, but we should pick one.

Reply:
(note: this was in reference to the `addAttribute` function but the Github UI is linking it to the `classNameAndIHaveSpokenToDST` for some reason.

Reply:
Renaming to `addDataAttribute`, or `dataAttribute`, seems like the best approach. Seems like the `add.*` convention should only be used when records are being merged / you can pipe multiple options into a record... what do you think?:
```
(Heading.h1
    |> Heading.variant Heading6
    |> Heading.addAttribute { name = "automation-id", value = "inbox-heading" }
    |> Heading.addAttribute { name = "another-attr", value = "yo" }
)
```

I think we should consider aria and data attributes separately. I don't feel overly strongly about this though, but here's my rationale: 

In Typescript, data attributes aren't valid JS identifiers so you can get away without having types, which is what I've done in the React component (https://www.typescriptlang.org/docs/handbook/jsx.html#attribute-type-checking), whereas aria attributes are typed (https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/types/react/index.d.ts#L1516). 

E.g.
```
export const Heading = ({
  classNameAndIHaveSpokenToDST,
  children,
  tag,
  variant,
  ...otherProps // this picks up data-* 
}: HeadingProps & AriaAttributes) // adds types for aria-* attributes
```
I'm feeling that the most consistent approach is to consider them separate for both Elm and React. In Elm we might also want to be more explicit about the types of aria attributes instead of accepting any key, something like this: https://package.elm-lang.org/packages/fapian/elm-html-aria/latest/

Thoughts?

Reply:
Yeah, I think it's fine to have separate params: dataAttributes, ariaAttributes, etc. Definitely having the ariaAttributes as a separate param makes it possible to use https://package.elm-lang.org/packages/fapian/elm-html-aria/latest/ (otherwise it'd have to be forked to make it work with a custom data type, as it's output type is Html.Attribute which is opaque).

On naming, my choice would be `addDataAttribute`. The `add` bit is to indicate that you can add more than one (I think you should be able to), unlike `id` or `variant` which can take only one value. `dataAttribute` sounds like you can only have one. Alternatively it could be `addDataAttributes` and take a list, but I think just having `addDataAttribute` is all that's necessary.

There's a couple of other places I noticed where the name is too generic. I'll add a comment on them.